src/server/pkg/stream/priority_queue.go

```diff
@@ -0,0 +1,136 @@
+package stream
+
+import (
+	"io"
+)
+
+// Stream is the standard interface for a sorted stream that can be used in a PriorityQueue.
+type Stream interface {
+	Next() error
+	Key() string
+	Priority() int
+}
+
+// PriorityQueue implements a priority queue that operates on streams.
+type PriorityQueue struct {
+	queue []Stream
+	size  int
+	ss    []Stream
+}
+
+// NewPriorityQueue creates a new priority queue.
+func NewPriorityQueue(ss []Stream) *PriorityQueue {
+	return &PriorityQueue{
+		queue: make([]Stream, len(ss)+1),
+		ss:    ss,
+	}
+}
+
+// Iterate iterates through the priority queue.
+func (pq *PriorityQueue) Iterate(cb func([]Stream, ...string) error) error {
```

Comment:
So, iterate does mean to do something repeatedly, which is what this is doing, but I kind of agree with you that `ForEach` seems like a better name here (we actually use `Iterate` throughout the storage layer for a similar thing, so I am also considering switching that over so we can be consistent across the board). `ForEach` seems to better describe this specific type of iteration because it is executing a callback for each step in the iteration.